### PR TITLE
manual update to py38-linux-64.lock (#4480)

### PR DESCRIPTION
**Pull request by bjlittle**
### Description
This PR manually updates the `py38-linux-64.lock` based on the contents of GHA PR #4480 (which I've now closed)

This PR includes the recent changes (#4486) to address supporting the latest `numpy` 1.22.0 release.

There are failures to investigate here, due to the dependency updates... so this PR is a work-in-progress, but also mainly for transparency given the current `iris` 3.2.0 release efforts.

Failures to address:
- `iris.tests.integration.test_trajectory`
  - [x] `TestTriPolar.test_tri_polar`
  - [x] `TestLazyData.test_hybrid_height`
- `iris.tests.unit.analysis.cartography.test_rotate_grid_vectors.TestRotateGridVectors`
  - [x] `test_angles_calculation`
  - [x] `test_angles_in_radians`
  - [x] `test_nan_vectors`
- `iris.tests.integration.plot.test_vector_plots.TestBarbs`
  - [x] `test_2d_plain_latlon`
  - [x] `test_2d_rotated_latlon`
  - [x] `test_2d_plain_latlon_on_polar_map`  

Reference GHA PR #4507 (closed)

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -41,6 +41,9 @@ env:
   IRIS_TEST_DATA_VERSION: "2.5"
   # Base directory for the iris-test-data.
   IRIS_TEST_DATA_DIR: ${HOME}/iris-test-data
+  # https://github.com/numpy/numpy/pull/19478
+  # https://github.com/matplotlib/matplotlib/pull/22099
+  NPY_DISABLE_CPU_FEATURES: "AVX512F,AVX512CD,AVX512VL,AVX512BW,AVX512DQ,AVX512_SKX"
```
Comment (bjlittle) [inline .cirrus.yml:46]: This is just horrible.

But it's necessary to disable this CPU feature to ensure that the tests running on Cloud CI platforms behave the same as local host runs; provided the local host doesn't also have Intel Short Vector Math Library (SVML) support for the platform chipset (in which case, also disable it)

File: .cirrus.yml [hunk 2]
```diff
@@ -60,7 +63,6 @@ linux_task_template: &LINUX_TASK_TEMPLATE
       - echo "$(date +%Y).$(expr $(date +%U) / ${CACHE_PERIOD}):${CONDA_CACHE_BUILD}"
       - uname -r
     populate_script:
-      - export CONDA_OVERRIDE_LINUX="$(uname -r | cut -d'+' -f1)"
```
Comment (bjlittle) [inline .cirrus.yml:63]: This isn't needed any more

File: lib/iris/tests/integration/test_trajectory.py [hunk 3]
```diff
@@ -216,8 +216,10 @@ def setUp(self):
         cube.coord("depth").bounds = b32
         self.cube = cube
         # define a latitude trajectory (put coords in a different order
-        # to the cube, just to be awkward)
-        latitudes = list(range(-90, 90, 2))
+        # to the cube, just to be awkward) although avoid south pole
+        # singularity as a sample point and the issue of snapping to
+        # multi-equidistant closest points from within orca antarctic hole
+        latitudes = list(range(-80, 90, 2))
```
Comment (bjlittle) [inline lib/iris/tests/integration/test_trajectory.py:222]: This test is an accident waiting to happen... let's just make it a wee bit more robust and avoid a known possible problem.

File: lib/iris/tests/integration/test_trajectory.py [hunk 4]
```diff
@@ -226,7 +228,9 @@ def setUp(self):
 
     def test_tri_polar(self):
         # extract
-        sampled_cube = traj_interpolate(self.cube, self.sample_points)
+        sampled_cube = traj_interpolate(
+            self.cube, self.sample_points, method="nearest"
+        )
```
Comment (bjlittle) [inline lib/iris/tests/integration/test_trajectory.py:233]: Just making the current behaviour explicit

File: lib/iris/tests/integration/test_trajectory.py [hunk 5]
```diff
@@ -329,8 +333,12 @@ def test_hybrid_height(self):
         # Put a lazy array into the cube so we can test deferred loading.
         cube.data = as_lazy_data(cube.data)
 
+        # Use opionated grid-latitudes to avoid the issue of platform
+        # specific behaviour within SciPy cKDTree choosing a different
+        # equi-distant nearest neighbour point when there are multiple
+        # valid candidates.
         traj = (
-            ("grid_latitude", [20.5, 21.5, 22.5, 23.5]),
+            ("grid_latitude", [20.4, 21.6, 22.6, 23.6]),
```
Comment (bjlittle) [inline lib/iris/tests/integration/test_trajectory.py:341]: We could leave this as-is... but I'm again erring on the side of doing the obvious for clarity and robustness sake.

File: lib/iris/tests/unit/analysis/cartography/test_rotate_grid_vectors.py [hunk 6]
```diff
@@ -33,7 +33,7 @@ def _check_angles_calculation(
         u_cube.rename("dx")
         u_cube.data[...] = 0
         v_cube = u_cube.copy()
-        v_cube.name("dy")
+        v_cube.rename("dy")
```
Comment (bjlittle) [inline lib/iris/tests/unit/analysis/cartography/test_rotate_grid_vectors.py:36]: Spotted this... so fixed it, but has no real consequence.

File: lib/iris/tests/unit/analysis/cartography/test_rotate_grid_vectors.py [hunk 7]
```diff
@@ -71,7 +71,7 @@ def _check_angles_calculation(
         ang_diffs = out_angs - expect_angs
         # Fix for null vectors, and +/-360 differences.
         ang_diffs[np.abs(out_mags) < 0.001] = 0.0
-        ang_diffs = ang_diffs % 360.0
+        ang_diffs[np.isclose(np.abs(ang_diffs), 360.0)] = 0.0
```
Comment (bjlittle) [inline lib/iris/tests/unit/analysis/cartography/test_rotate_grid_vectors.py:74]: Same result, but more robust given the floating point numbers we're dealing with.

**Comment by bjlittle** [comment]
Additionally, we could choose to additionally pass the `rounding=False` kwarg to the following three `barbs` tests:
- `iris.tests.integration.plot.test_vector_plots.TestBarbs`
  - `test_2d_plain_latlon`
  - `test_2d_rotated_latlon`
  - `test_2d_plain_latlon_on_polar_map`  

This will counter the disparity in behaviour between Intel SVML supported and unsupported CPUs, but this would also mean updating the associated `imagerepo.json` entries and adding additional images to the `test-iris-imagehash` repo.

I'm not fussed either way, but it would have the advantage of a user with a local host with SVML support being able to run tests that pass independent of disabling the `numpy` AVX512...

File: .cirrus.yml (hunk 1 above)
Comment (bjlittle) [reply .cirrus.yml:46]: Might be best to push this down into the `noxfile` instead...

Comment (bjlittle) [reply .cirrus.yml]: Done.

**Review by jamesp — APPROVED** [review]

**Comment by bjlittle** [comment]
Awesome, thanks @jamesp :beers:

**Pull request merged**